File: essential-code-backup/apps/backend-ts/src/websocket.ts
/**
 * WebSocket Server for Real-Time Features
 *
 * Bridges Redis pub/sub to WebSocket clients
 * Enables real-time notifications without polling
 */

import { Server, Socket } from 'socket.io';
import { Server as HTTPServer } from 'http';
import logger from './services/logger.js';
import { PubSubService, CHANNELS, UserNotification, AIResult } from './utils/pubsub.js';

export function setupWebSocket(httpServer: HTTPServer) {
  const io = new Server(httpServer, {
    cors: {
      origin: process.env.WEBAPP_URL || 'https://zantara.balizero.com',
      methods: ['GET', 'POST'],
      credentials: true,
    },
    transports: ['websocket', 'polling'],
  });

  logger.info('WebSocket server initializing...');

  // Authentication middleware
  io.use((socket, next) => {
    const userId = socket.handshake.auth.userId;
    if (!userId) {
      return next(new Error('Authentication required'));
    }
    socket.data.userId = userId;
    next();
  });

  // Connection handling
  io.on('connection', (socket: Socket) => {
    const userId = socket.data.userId;
    logger.info(`User ${userId} connected via WebSocket`);

    // Join user-specific room
    socket.join(`user:${userId}`);

    // Send welcome message
    socket.emit('connected', {
      message: 'Real-time connection established',
      userId,
      timestamp: Date.now(),
    });

    // Handle disconnection
    socket.on('disconnect', (reason) => {
      logger.info(`User ${userId} disconnected: ${reason}`);
    });

    // Handle ping (keep-alive)
    socket.on('ping', () => {
      socket.emit('pong', { timestamp: Date.now() });
    });

    // Handle room joining (for chat)
    socket.on('join-room', (roomId: string) => {
      socket.join(`room:${roomId}`);
      logger.info(`User ${userId} joined room ${roomId}`);
      socket.emit('room-joined', { roomId });
    });

    // Handle room leaving
    socket.on('leave-room', (roomId: string) => {
      socket.leave(`room:${roomId}`);
      logger.info(`User ${userId} left room ${roomId}`);
    });
  });

  // Bridge Redis pub/sub to WebSocket
  setupRedisBridge(io);

  logger.info('✅ WebSocket server ready');

  return io;
}

/**
 * Bridge Redis pub/sub events to WebSocket clients
 */
function setupRedisBridge(io: Server) {
  // User notifications
  PubSubService.psubscribe<UserNotification>(
    `${CHANNELS.USER_NOTIFICATIONS}:*`,
    (_channel, notification) => {
      const userId = notification.userId;
      io.to(`user:${userId}`).emit('notification', notification);
      logger.debug(`Notification sent to user ${userId}`);
    }
  );

  // AI results
  PubSubService.psubscribe<AIResult>(`${CHANNELS.AI_RESULTS}:*`, (_channel, result) => {
    const userId = result.userId;
    io.to(`user:${userId}`).emit('ai-result', result);
    logger.debug(`AI result sent to user ${userId}`);
  });

  // Chat messages
  PubSubService.psubscribe(`${CHANNELS.CHAT_MESSAGES}:*`, (channel, message: any) => {
    const roomId = channel.split(':')[2];
    io.to(`room:${roomId}`).emit('chat-message', message);
    logger.debug(`Chat message sent to room ${roomId}`);
  });

  // System events (broadcast to all)
  PubSubService.subscribe(CHANNELS.SYSTEM_EVENTS, (event: any) => {
    io.emit('system-event', event);
    logger.debug('System event broadcasted');
  });

  logger.info('✅ Redis → WebSocket bridge established');
}

/**
 * Graceful shutdown
 */
export async function closeWebSocket(io: Server): Promise<void> {
  return new Promise((resolve) => {
    io.close(() => {
      logger.info('WebSocket server closed');
      resolve();
    });
  });
}

File: apps/backend-rag/backend/app/routers/websocket.py
"""
WebSocket Router & Connection Manager
Handles real-time notifications via Redis Pub/Sub
Replaces Node.js websocket.ts
"""

import asyncio
import json
import logging
from typing import Any

import redis.asyncio as redis
from fastapi import APIRouter, WebSocket, WebSocketDisconnect, Query, Depends
from jose import jwt, JWTError

from app.core.config import settings

logger = logging.getLogger(__name__)

router = APIRouter(tags=["websocket"])


class ConnectionManager:
    """
    Manages active WebSocket connections.
    """

    def __init__(self):
        # Map user_id -> List[WebSocket]
        self.active_connections: dict[str, list[WebSocket]] = {}
        self.lock = asyncio.Lock()

    async def connect(self, websocket: WebSocket, user_id: str):
        await websocket.accept()
        async with self.lock:
            if user_id not in self.active_connections:
                self.active_connections[user_id] = []
            self.active_connections[user_id].append(websocket)
            logger.info(f"🔌 WebSocket connected: {user_id} (Total: {len(self.active_connections[user_id])})")

    async def disconnect(self, websocket: WebSocket, user_id: str):
        async with self.lock:
            if user_id in self.active_connections:
                if websocket in self.active_connections[user_id]:
                    self.active_connections[user_id].remove(websocket)
                if not self.active_connections[user_id]:
                    del self.active_connections[user_id]
        logger.info(f"🔌 WebSocket disconnected: {user_id}")

    async def send_personal_message(self, message: dict, user_id: str):
        """Send message to all connections of a specific user"""
        if user_id in self.active_connections:
            # Create a copy of the list to avoid modification during iteration issues
            connections = self.active_connections[user_id][:]
            for connection in connections:
                try:
                    await connection.send_json(message)
                except Exception as e:
                    logger.warning(f"⚠️ Failed to send WS message to {user_id}: {e}")
                    # Cleanup dead connection
                    await self.disconnect(connection, user_id)

    async def broadcast(self, message: dict):
        """Broadcast message to ALL connected users"""
        # Iterate over a copy of keys
        user_ids = list(self.active_connections.keys())
        for user_id in user_ids:
            await self.send_personal_message(message, user_id)


manager = ConnectionManager()


async def get_current_user_ws(
    token: str = Query(...)
) -> str | None:
    """
    Validate JWT token from query parameter
    """
    try:
        payload = jwt.decode(
            token, settings.jwt_secret_key, algorithms=[settings.jwt_algorithm]
        )
        user_id: str = payload.get("sub") or payload.get("userId")
        if user_id is None:
            return None
        return user_id
    except JWTError:
        return None


@router.websocket("/ws")
async def websocket_endpoint(
    websocket: WebSocket,
    token: str = Query(...)
):
    """
    WebSocket endpoint for real-time updates.
    Requires valid JWT token in query param: /ws?token=...
    """
    user_id = await get_current_user_ws(token)
    
    if not user_id:
        logger.warning("⚠️ WebSocket connection rejected: Invalid token")
        await websocket.close(code=4003) # Forbidden
        return

    await manager.connect(websocket, user_id)

    try:
        while True:
            # Keep connection alive and listen for client messages (if any)
            # Currently we primarily push server->client, but client might send pings
            data = await websocket.receive_text()
            # Optional: Handle client messages here (e.g. ping/pong)
            
    except WebSocketDisconnect:
        await manager.disconnect(websocket, user_id)
    except Exception as e:
        logger.error(f"❌ WebSocket error: {e}")
        await manager.disconnect(websocket, user_id)


# ============================================================================
# Redis Pub/Sub Listener
# ============================================================================

async def redis_listener():
    """
    Background task to listen for Redis Pub/Sub events and forward to WebSockets
    """
    if not settings.redis_url:
        logger.warning("⚠️ Redis URL not set. WebSocket notifications disabled.")
        return

    redis_client = redis.from_url(settings.redis_url, decode_responses=True)
    pubsub = redis_client.pubsub()

    # Subscribe to channels
    # Patterns:
    # CHANNELS.USER_NOTIFICATIONS:* -> user specific
    # CHANNELS.AI_RESULTS:* -> user specific (usually)
    # CHANNELS.CHAT_MESSAGES:* -> room/user specific
    # CHANNELS.SYSTEM_EVENTS -> broadcast
    
    # We subscribe to patterns to catch all
    await pubsub.psubscribe("CHANNELS.USER_NOTIFICATIONS:*")
    await pubsub.psubscribe("CHANNELS.AI_RESULTS:*")
    await pubsub.psubscribe("CHANNELS.CHAT_MESSAGES:*")
    await pubsub.subscribe("CHANNELS.SYSTEM_EVENTS")

    logger.info("✅ Redis Pub/Sub listener started")

    try:
        async for message in pubsub.listen():
            if message["type"] in ["message", "pmessage"]:
                channel = message["channel"]
                data_str = message["data"]
                
                try:
                    data = json.loads(data_str)
                except json.JSONDecodeError:
                    data = {"raw_data": data_str}

                # Determine target user from channel name
                # Format: CHANNELS.USER_NOTIFICATIONS:{userId}
                
                if "USER_NOTIFICATIONS" in channel:
                    # Extract userId
                    parts = channel.split(":")
                    if len(parts) > 1:
                        target_user_id = parts[-1]
                        await manager.send_personal_message(
                            {"type": "notification", "data": data}, 
                            target_user_id
                        )
                
                elif "AI_RESULTS" in channel:
                    parts = channel.split(":")
                    if len(parts) > 1:
                        target_user_id = parts[-1]
                        await manager.send_personal_message(
                            {"type": "ai-result", "data": data}, 
                            target_user_id
                        )
                
                elif "CHAT_MESSAGES" in channel:
                    # Chat messages might be for a room or user. 
                    # For now, assuming direct mapping or we broadcast to room members.
                    # Simplified: If channel has userId, send to user.
                    parts = channel.split(":")
                    if len(parts) > 1:
                        target_id = parts[-1]
                        # Try sending to user (if target is user)
                        await manager.send_personal_message(
                            {"type": "chat-message", "data": data}, 
                            target_id
                        )
                        # If target is room, we'd need room management logic (TODO)
                
                elif "SYSTEM_EVENTS" in channel:
                    await manager.broadcast({"type": "system-event", "data": data})

    except asyncio.CancelledError:
        logger.info("🛑 Redis listener cancelled")
    except Exception as e:
        logger.error(f"❌ Redis listener error: {e}")
    finally:
        await pubsub.close()
        await redis_client.close()
